pkg/util/fileconfig.go

```diff
@@ -0,0 +1,257 @@
+/*
+Copyright 2019 ceph-csi authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+    "encoding/json"
+    "fmt"
+    "io/ioutil"
+    "strings"
+)
+
+/* FileConfig processes config information stored in files, mostly mapped into
+   the runtime container.
+
+   The calls explicitly do not cache any information, to ensure that updated
+   configuration is always read from the files (for example when these are
+   mapped in as k8s config maps or secrets).
+
+   The BasePath is the path where config files are found, and config files are
+   expected to be named in the following manner,
+    - BasePath/ceph-cluster-<cluster-fsid>/cluster-config
+    - BasePath/ceph-cluster-<cluster-fsid>-provisioner-secret/credentials
+    - BasePath/ceph-cluster-<cluster-fsid>-provisioner-secret/subjectid
+    - BasePath/ceph-cluster-<cluster-fsid>-publish-secret/credentials
+    - BasePath/ceph-cluster-<cluster-fsid>-publish-secret/subjectid
+    Where,
+        - cluster-fsid is the Ceph cluster fsid in UUID ascii notation
+        - The cluster-fsid corresponds to the cluster for which the
+        configuration information is present in the mentioned files
+        - cluster-config is expected to be a JSON blob with the following
+        structure,
+        {
+            "version": 1,
+            "cluster-config": {
+                "cluster-fsid": "<ceph-fsid>",
+                "monitors": [
+                    "IP/DNS:port",
+                    "IP/DNS:port"
+                ],
+                "pools": [
+                    "<pool-name>",
+                    "<pool-name>"
+                ]
+            }
+        }
+        - credentials is expected to contain Base64 encoded credentials for the
+        user encoded in subjectid
+        - subjectid is the username/subject to use with calls to Ceph, and is
+        also Base64 encoded
+        - Provisioner secret contains secrets to use by the provisioning system
+        - Publish secret contains secrets to use by the publishing/staging
+        system
+*/
+
+// FileConfig type with basepath that points to source of all config files
+type FileConfig struct {
+    BasePath string
+}
+
+// ClusterConfigv1 strongly typed JSON spec for cluster-config above
+type ClusterConfigv1 struct {
```

Comment:
question: why  the same v1?

Reply:
Removed in the current iteration. The reasoning was, we can change the data in the JSON bloc, and hence it had a version, and that maps to versioned structures, and hence the name. All this is moot for the present though, as these are removed for now.